[web-animations] "Commits transforms as a transform list" subtest fails in web-animations/interfaces/Animation/commitStyles.html

graouts opened the pull request:
#### 7192e46653d27456404abdb72ca60bb945279d68
<pre>
[web-animations] &quot;Commits transforms as a transform list&quot; subtest fails in web-animations/interfaces/Animation/commitStyles.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=246011">https://bugs.webkit.org/show_bug.cgi?id=246011</a>
rdar://101021972

Reviewed by Martin Robinson.

We should never attempt to use the matrix-based blending when either transform list is empty.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animation/commitStyles-expected.txt:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
(WebCore::TransformOperations::blend const):

Canonical link: <a href="https://commits.webkit.org/255946@main">https://commits.webkit.org/255946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4394e4059cba1a22268b4553fc13d5f29c9db63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3053 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24435 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103493 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163827 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3070 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31288 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86181 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99499 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99528 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2191 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80285 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29213 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84117 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83818 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72165 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37684 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17647 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35548 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18910 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41480 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1933 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38140 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mrobinson commented on Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
```diff
@@ -106,7 +106,7 @@ TransformOperations TransformOperations::blend(const TransformOperations& from,
         RefPtr<TransformOperation> fromOperation = (i < fromOperationCount) ? from.operations()[i].get() : nullptr;
         RefPtr<TransformOperation> toOperation = (i < toOperationCount) ? operations()[i].get() : nullptr;
-        if ((prefixLength && i >= *prefixLength) || (fromOperation && toOperation && !fromOperation->sharedPrimitiveType(toOperation.get()))) {
+        if ((prefixLength && i >= *prefixLength && *prefixLength) || (fromOperation && toOperation && !fromOperation->sharedPrimitiveType(toOperation.get()))) {
```
This seems wrong to me. There are other cases where the prefix might be empty as well such as well the operation lists do not match at all. I think in this case where one list is empty the prefix should be as long as the non-empty list and have compatible primitives. The code that handles this is in `SharedPrimitivesPrefix::update`:

```c++
        // If we haven't seen an operation at this index before, we can simply use our primitive type.
        if (i >= m_primitives.size()) {
            ASSERT(i == m_primitives.size());
            m_primitives.append(operation->primitiveType());
            continue;
        }
```

I suspect something else is going on and this might just be covering up the original issue.

graouts replied:
In the affected test, `SharedPrimitivesPrefix::update()` is never called due to the early return in `KeyframeEffect::checkForMatchingTransformFunctionLists()`:

```c++
void KeyframeEffect::checkForMatchingTransformFunctionLists()
{
    if (m_blendingKeyframes.size() < 2 || !m_blendingKeyframes.containsProperty(CSSPropertyTransform)) {
        m_transformFunctionListsMatchPrefix = 0;
        return;
    }

    SharedPrimitivesPrefix prefix;
    for (const auto& keyframe : m_blendingKeyframes)
        prefix.update(keyframe.style()->transform());

    m_transformFunctionListsMatchPrefix = prefix.primitives().size();
}
```

Indeed, in this case, there is a single keyframe.

graouts replied:
So my guess was to remove the `m_blendingKeyframes.size() < 2` clause which would only set `m_transformFunctionListsMatchPrefix` to `0` should there not be a single keyframe specifying a `transform`. This lets `SharedPrimitivesPrefix::update()` run and set the value to `1`.

But this regresses at least one test, `compositing/backing/backing-store-attachment-empty-keyframe.html`, which no longer produces a layer tree.

graouts replied:
What seems to me like a simple approach is to never attempt to do matrix-blending when either the `from` or `to` transform list is empty, not involving the prefix at all.

mrobinson replied:
This version of the change does look more reasonable, but I wonder...do things still work well if you simply change the line in question to this:

```c++
    if (m_blendingKeyframes.size() < 1 || !m_blendingKeyframes.containsProperty(CSSPropertyTransform)) {
        m_transformFunctionListsMatchPrefix = 0;
        return;
    }
```

graouts replied:
Are you suggesting to make that change in `KeyframeEffect::checkForMatchingTransformFunctionLists()` exclusively or on top of my current patch? I suppose the latter, right?

graouts replied:
Also, if we don't have _any_ blending keyframes, then we won't blend at all, so the `m_transformFunctionListsMatchPrefix` value is really irrelevant.

graouts replied:
That said, to answer your question, your proposed change as an additional change on top of the current version of this patch regresses `compositing/backing/backing-store-attachment-empty-keyframe.html`.

mrobinson reviewed (approved):
Sorry. I've read this again a bit more carefully. This looks great to me.

graouts committed:
[web-animations] "Commits transforms as a transform list" subtest fails in web-animations/interfaces/Animation/commitStyles.html
https://bugs.webkit.org/show_bug.cgi?id=246011
rdar://101021972

Reviewed by Martin Robinson.

We should never attempt to use the matrix-based blending when either transform list is empty.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animation/commitStyles-expected.txt:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
(WebCore::TransformOperations::blend const):

Canonical link: https://commits.webkit.org/255946@main
